File: Code/Plugins/Revenj.Plugins.Server.Commands/ExecuteService.cs
using System;
using System.Collections.Concurrent;
using System.ComponentModel.Composition;
using System.Diagnostics.Contracts;
using System.Linq;
using System.Net;
using System.Runtime.Serialization;
using NGS;
using NGS.Common;
using NGS.DomainPatterns;
using NGS.Extensibility;
using NGS.Security;
using NGS.Serialization;
using NGS.Utility;
using Revenj.Processing;

namespace Revenj.Plugins.Server.Commands
{
	[Export(typeof(IServerCommand))]
	[ExportMetadata(Metadata.ClassType, typeof(ExecuteService))]
	public class ExecuteService : IServerCommand
	{
		private readonly IServiceLocator Locator;
		private readonly ITypeResolver TypeResolver;
		private readonly IPermissionManager Permissions;

		public ExecuteService(
			IServiceLocator locator,
			ITypeResolver typeResolver,
			IPermissionManager permissions)
		{
			Contract.Requires(locator != null);
			Contract.Requires(typeResolver != null);
			Contract.Requires(permissions != null);

			this.Locator = locator;
			this.TypeResolver = typeResolver;
			this.Permissions = permissions;
		}

		[DataContract(Namespace = "")]
		public class Argument<TFormat>
		{
			[DataMember]
			public string Name;
			[DataMember]
			public TFormat Data;
		}

		private static TFormat CreateExampleArgument<TFormat>(ISerialization<TFormat> serializer)
		{
			return serializer.Serialize(new Argument<TFormat> { Name = "CheckInfo" });
		}

		private static ConcurrentDictionary<Type, IExecuteCommand> Cache = new ConcurrentDictionary<Type, IExecuteCommand>();

		public ICommandResult<TOutput> Execute<TInput, TOutput>(ISerialization<TInput> input, ISerialization<TOutput> output, TInput data)
		{
			var either = CommandResult<TOutput>.Check<Argument<TInput>, TInput>(input, output, data, CreateExampleArgument);
			if (either.Error != null)
				return either.Error;
			var argument = either.Argument;

			var serviceType = TypeResolver.Resolve(argument.Name);
			if (serviceType == null)
				return
					CommandResult<TOutput>.Fail(
						"Couldn't find service: {0}".With(argument.Name),
						@"Example argument: 
" + CommandResult<TOutput>.ConvertToString(CreateExampleArgument(output)));

			var serviceInterface =
				serviceType.GetInterfaces()
				.FirstOrDefault(it => it.IsGenericType && typeof(IServerService<,>) == it.GetGenericTypeDefinition());
			if (serviceInterface == null)
				return
					CommandResult<TOutput>.Fail(
						"Object: {0} is not a valid service.".With(argument.Name),
						"{0} must implement {1} to be executed as a service call".With(argument.Name, typeof(IServerService<,>).FullName));

			if (!Permissions.CanAccess(serviceType))
				return
					CommandResult<TOutput>.Return(
						HttpStatusCode.Forbidden,
						default(TOutput),
						"You don't have permission to access: {0}.",
						argument.Name);

			try
			{
				IExecuteCommand command;
				if (!Cache.TryGetValue(serviceType, out command))
				{
					var commandType = typeof(ExecuteServiceCommand<,>).MakeGenericType(serviceInterface.GetGenericArguments());
					command = Activator.CreateInstance(commandType) as IExecuteCommand;
					Cache.TryAdd(serviceType, command);
				}
				var result = command.Execute(input, output, Locator, serviceType, argument.Data);

				return CommandResult<TOutput>.Return(HttpStatusCode.Created, result, "Service executed");
			}
			catch (ArgumentException ex)
			{
				return CommandResult<TOutput>.Fail(
					ex.Message,
					ex.GetDetailedExplanation() + @"
Example argument: 
" + CommandResult<TOutput>.ConvertToString(CreateExampleArgument(output)));
			}
		}

		private interface IExecuteCommand
		{
			TOutput Execute<TInput, TOutput>(
				ISerialization<TInput> input,
				ISerialization<TOutput> output,
				IServiceLocator locator,
				Type serviceType,
				TInput data);
		}

		private class ExecuteServiceCommand<TArgument, TResult> : IExecuteCommand
		{
			public TOutput Execute<TInput, TOutput>(
				ISerialization<TInput> input,
				ISerialization<TOutput> output,
				IServiceLocator locator,
				Type serviceType,
				TInput data)
			{
				TArgument arg;
				try
				{
					arg = data != null ? input.Deserialize<TInput, TArgument>(data, locator) : Activator.CreateInstance<TArgument>();
				}
				catch (Exception ex)
				{
					throw new ArgumentException("Error deserializing service argument.", ex);
				}
				IServerService<TArgument, TResult> service;
				try
				{
					service = locator.Resolve<IServerService<TArgument, TResult>>(serviceType);
				}
				catch (Exception ex)
				{
					throw new ArgumentException("Can't create service instance.", ex);
				}
				try
				{
					return output.Serialize(service.Execute(arg));
				}
				catch (Exception ex)
				{
					string additionalInfo;
					try
					{
						additionalInfo = @"Sent data:
" + input.Serialize(arg);
					}
					catch (Exception sex)
					{
						additionalInfo = "Error serializing input: " + sex.Message;
					}
					throw new ArgumentException(
						ex.Message,
						new FrameworkException(@"Error while executing service: {0}. {1}".With(ex.Message, additionalInfo), ex));
				}
			}
		}
	}
}

File: java/revenj-servlet/src/main/java/org/revenj/server/commands/ExecuteService.java
package org.revenj.server.commands;

import org.revenj.patterns.DomainModel;
import org.revenj.patterns.SearchableRepository;
import org.revenj.patterns.ServiceLocator;
import org.revenj.patterns.Specification;
import org.revenj.security.PermissionManager;
import org.revenj.serialization.Serialization;
import org.revenj.server.CommandResult;
import org.revenj.server.ServerCommand;
import org.revenj.server.ServerService;

import java.io.IOException;
import java.lang.reflect.Constructor;
import java.lang.reflect.Method;
import java.lang.reflect.Type;
import java.security.Principal;
import java.util.Arrays;
import java.util.Optional;

public class ExecuteService implements ServerCommand {

	private final ClassLoader loader;
	private final PermissionManager permissions;

	public ExecuteService(
			ClassLoader loader,
			PermissionManager permissions) {
		this.loader = loader;
		this.permissions = permissions;
	}

	public static final class Argument<TFormat> {
		public String Name;
		public TFormat Data;

		public Argument(String name, TFormat data) {
			this.Name = name;
			this.Data = data;
		}

		@SuppressWarnings("unused")
		private Argument() {
		}
	}

	@Override
	public <TInput, TOutput> CommandResult<TOutput> execute(
			ServiceLocator locator,
			Serialization<TInput> input,
			Serialization<TOutput> output,
			TInput data,
			Principal principal) {
		Argument<TInput> arg;
		try {
			arg = input.deserialize(data, Argument.class, data.getClass());
		} catch (IOException e) {
			return CommandResult.badRequest(e.getMessage());
		}
		Class<?> manifest;
		try {
			manifest = Class.forName(arg.Name, true, loader);
		} catch (ClassNotFoundException e) {
			return CommandResult.badRequest("Unable to find specified service: " + arg.Name);
		}
		if (!ServerService.class.isAssignableFrom(manifest)) {
			return CommandResult.badRequest("Object: " + arg.Name + " is not a valid service.");
		}
		Optional<Method> method =
				Arrays.stream(manifest.getDeclaredMethods())
						.filter(it -> "execute".equals(it.getName()) && it.getGenericParameterTypes().length == 1)
						.findFirst();
		if (!method.isPresent()) {
			return CommandResult.badRequest("Object: " + arg.Name + " is not a valid service.");
		}
		if (!permissions.canAccess(manifest, principal)) {
			return CommandResult.forbidden(arg.Name);
		}
		Class<?> argumentType = method.get().getParameterTypes()[0];
		Object argument;
		try {
			argument = input.deserialize(arg.Data, argumentType);
		} catch (IOException e) {
			return CommandResult.badRequest(e.getMessage());
		}
		ServerService service = null;
		try {
			service = (ServerService) locator.resolve((Type) manifest);
		} catch (ReflectiveOperationException e) {
			for (Constructor ctor : manifest.getConstructors()) {
				try {
					service = (ServerService) locator.create(ctor);
				} catch (ReflectiveOperationException ignore) {
				}
			}
			if (service == null) {
				return new CommandResult<>(null, "Unable to create an instance of: " + manifest, 500);
			}
		}
		Object result = service.execute(argument);
		try {
			return CommandResult.success("Service executed", output.serialize(result));
		} catch (IOException e) {
			return new CommandResult<>(null, "Error serializing result.", 500);
		}
	}
}
